File: code_samples/typescript/oop_abstraction.ts
/**
 * Oop: Abstraction
 * AI/ML Training Sample
 */

interface IAbstraction {
    data: string;
    process(input: string): void;
    validate(): boolean;
}

class Abstraction implements IAbstraction {
    data: string;
    
    constructor() {
        this.data = "";
    }
    
    process(input: string): void {
        this.data = input;
    }
    
    getData(): string {
        return this.data;
    }
    
    validate(): boolean {
        return this.data.length > 0;
    }
}

// Example usage
const instance = new Abstraction();
instance.process("example");
console.log(`Data: ${instance.getData()}`);
console.log(`Valid: ${instance.validate()}`);

export { Abstraction, IAbstraction };

File: code_samples/elixir/oop_abstraction.ex
# Oop: Abstraction
# AI/ML Training Sample

defmodule Abstraction do
  defstruct data: ""
  
  def new(), do: %Abstraction{}
  
  def process(%Abstraction{} = struct, input) do
    %{struct | data: input}
  end
  
  def get_data(%Abstraction{data: data}), do: data
  
  def validate(%Abstraction{data: data}) do
    String.length(data) > 0
  end
end

# Example usage
instance = Abstraction.new()
updated = Abstraction.process(instance, "example")
IO.puts("Data: " <> Abstraction.get_data(updated))
IO.puts("Valid: " <> to_string(Abstraction.validate(updated)))
